### ISSUE TITLE
Performance is very slow when multiple LinuxKit instance running simultaneously

### ISSUE THREAD

**Issue by arm64b** [linuxkit/linuxkit#2636]
**Description**

**Steps to reproduce the issue:**
1. open 2 ssh terminal to connect the same remote arm64 server, then
2. run 2 linuxkit instance at the same time in the above terminal with below command for each one:
terminal a: `bin/linuxkit run qemu -uefi -state "mysql" -fw /home/dennis/qemu/bios.bin -networking "tap,tap0" -iso arm64-efi.iso`
terminal b. `bin/linuxkit run qemu -uefi -state "apache" -fw /home/dennis/qemu/bios.bin -networking "tap,tap1" -iso arm64-efi.iso`

**Describe the results you received:**
We find that the LinuxKit will boot in a very slow speed comparing only one instance launched.

**Describe the results you expected:**
We hope the 2 LinuxKit instances will run as the same speed as we run only one LinuxKit instance each time.

**Additional information you deem important (e.g. issue happens only occasionally):**
I use `perf record` to record the performance related data for the entire life cycle of a LinuxKit instance(from `linuxkit run` to `poweroff` the LinuxKit system , the data as below (see the attached screen snapshot):
![linuxkit](https://user-images.githubusercontent.com/29669302/31983510-1dc151ce-b990-11e7-9ae6-e1df70303cc0.PNG)

**Comment by rn** [comment]
I can not repro this. I've booted two linukit instances, one after the other, using the `minimal.yml` using `qemu/kvm` on a arm64 server and both boot in about 4s.

In any case this is unlikely a LinuxKIt issue, but more likely an issue either with your system, the version of KVM or qemu.

I used:
```
[master]linuxkit-dev-arm64:t/linuxkit/examples: qemu-system-aarch64 --version
QEMU emulator version 2.5.0 (Debian 1:2.5+dfsg-5ubuntu10.16), Copyright (c) 2003-2008 Fabrice Bellard
[master]linuxkit-dev-arm64:t/linuxkit/examples: uname -a
Linux linuxkit-dev-arm64 4.10.0-26-generic #30~16.04.1-Ubuntu SMP Tue Jun 27 09:40:29 UTC 2017 aarch64 aarch64 aarch64 GNU/Linux
```

**Comment by arm64b** [comment]
I can reproduce this issue on both Qualcomm and Huawei arm64 server.
Rename the linuxkit.yml to arm64.yml, then:
`#bin/moby build -format iso-efi arm64.yml`
in terminal A, input the below command:
`#bin/linuxkit run qemu -uefi -state "apache" -fw "/home/dennis/qemu/images/bios.bin" -iso arm64-efi.iso`
in terminal B, input the below command:
`#bin/linuxkit run qemu -uefi -state "mysql" -fw "/home/dennis/qemu/images/bios.bin" -iso arm64-efi.iso`

Then the 2 terminals will bootup the LinuxKit in a very very slow speed (I even suspend the LinuxKit running on terminal 2 is hang for the slow speed :(

**Comment by arm64b** [comment]
`root@entos-huawei-d05:/home/dennis/container/linuxkit# uname -a
Linux entos-huawei-d05 4.9.20 #2 SMP Wed Jun 14 16:31:49 CST 2017 aarch64 aarch64 aarch64 GNU/Linux`

`root@entos-huawei-d05:/home/dennis/container/linuxkit# qemu-system-aarch64 --version
QEMU emulator version 2.9.50 (v2.9.0-1376-gcef8fd6-dirty)
Copyright (c) 2003-2017 Fabrice Bellard and the QEMU Project developers`

**Comment by rn** [comment]
Have you tried this with non-Linuxkit VMs? as said, this doesn't look like a LinuxKit issue to me

**Comment by arm64b** [comment]
Sorry for the late response. I've tested it on x86_64 with the same steps, this issue doesn't show on x86_64,
I am trying to take a look at what the root cause is...

**Comment by arm64b** [comment]
> Have you tried this with non-Linuxkit VMs? 

I tried to boot up a Centos VM image with QEMU on arm64 directly, but the QEMU will not allowed to launch the second Centos VM image with below error message:
```
root@entos-huawei-d05:/home/dennis/qemu/images# qemu-system-aarch64 -M virt -m 2048 -nographic -enable-kvm -cpu host -bios QEMU_EFI.fd.ok -hda CentOS-7-aarch64-64G-OK.qcow2
qemu-system-aarch64: -hda CentOS-7-aarch64-64G-OK.qcow2: Failed to get "write" lock
```

After googled, this seems related to:
[https://bugzilla.redhat.com/show_bug.cgi?id=1484130](https://bugzilla.redhat.com/show_bug.cgi?id=1484130)
[https://bugzilla.redhat.com/show_bug.cgi?id=1484130](https://bugzilla.redhat.com/show_bug.cgi?id=1378241)

The rational here is to protect the underlying VM image from destroying by multiple qemu instance, but this also stop some useful usage scenario to happen, just as mentioned in the 2nd link. LinuxKit doesn't have this issue since we don't save/change the LinuxKit image, at least for now.

**Comment by justincormack** [comment]
It appears there may be a --force-share qemu option to disable this.

**Comment by rn** [comment]
or just copy the diskimage and boot that

**Comment by arm64b** [comment]
> It appears there may be a --force-share qemu option to disable this.

Currently qemu doesn't have '--force-share' or similar option to disable this, seems this image share protection only for the qcow2 format image, for others like 'iso', the image share protection disabled automatically meaning for an iso image, it's possible to start multiple VM instances.

related discussion: [https://lists.gnu.org/archive/html/qemu-devel/2017-06/msg03732.html](https://lists.gnu.org/archive/html/qemu-devel/2017-06/msg03732.html)

**Comment by arm64b** [comment]
> or just copy the diskimage and boot that

I copy the arm64-efi.iso to arm64-efi-1.iso, then `linuxkit run` the 2 iso images individually, the issue is still there: one terminal has boot to the shell, another is still in the below output and will be into shell after, probably, several minutes will extremely slow speed:
`EFI stub: Booting Linux Kernel...`

**Comment by arm64b** [comment]
The issue is not only for iso format, it's also for the `-kernel ... -initrd...`. I haven't tried other formats...

**Comment by rn** [comment]
I can't repro this on the above system. I tried with `kernel+initrd` and:
```
sudo ~/tools/bin/linuxkit run qemu -kvm -state ./foo -networking "tap,tap0" minimal
sudo ~/tools/bin/linuxkit run qemu -kvm -state ./bar -networking "tap,tap1" minimal
```
in separate windows.

I can't see how this is an issue with LinuxKit itself and strongly suspect a KVM issue, so I'm closing this issue for now.

Having said that we can/should continue the discussion here.

@arm64b can you maybe try and update the kernel to something more recent. I seem to recall quite a few KVM arm64 changes flying by in various kernel updates, though I hadn't paid close attention

**Comment by arm64b** [comment]
@rn Two points:
1. You don't specify the firmware for the VM
2. if no FW specified, any image that depends on the boot load will not work.

So you can't boot a ISO image without UEFI FW. Please try to append a UEFI FW to your command like this:
`#bin/linuxkit run qemu -uefi -fw "/path/to/your/UEFI.fd" ...`

**Comment by rn** [comment]
yes, but you also said:
> The issue is not only for iso format, it's also for the -kernel ... -initrd...

and there you don't need firmware.

**Comment by arm64b** [comment]
The issue is not only for iso format, it's also for the -kernel...-initrd in case of UEFI FW specified, but '-kernel...-initrd...' doesn't have boot loader depends on UEFI, so you can boot the system without the UEFI, while an image with UEFI specific boot loader like efi-iso can't boot.

**Comment by rn** [comment]
why would you use UEFI if you boot `kernel+initrd`?

In any case i tried the same with iso-efi. It was a bit of a drag to get the FW, but (on ubuntu):
```
sudo apt-get install qemu-efi
dd if=/dev/zero of=flash0.img bs=1M count=64
dd if=/usr/share/qemu-efi/QEMU_EFI.fd of=flash0.img conv=notrunc
```
and then:
```
linuxkit run -iso -uefi -fw ~/flash0.img -state foo minimal-efi.iso
```
and at the sameish time in another window
```
linuxkit run -iso -uefi -fw ~/flash0.img -state bar minimal-efi.is
```
and both boot in 1.5 seconds.

**Comment by arm64b** [comment]
Hello @rn, I have tested on 3 platforms: Cavium ThunderX, Qualcomm Centriq 2400 and Huawei D05.
The issue doesn't show on Cavium platform, but I reproduce this issue on both Qualcomm and Huawei, even I upgrade the kernel to 4.13.11 (the latest stable version), we still can observe this scenario.

Which platform are you using?

**Comment by rn** [comment]
I only have access to Cavium Thunder (via packet).

**Comment by deitch** [comment]
@arm64b do you have access to a clean Qualcomm or Huawei on which we can test?

**Comment by arm64b** [comment]
Ah, both Qualcomm and Huawei are the local servers in our lab :) So I can't figure out the way you guys can access those 2 platforms. Can I record a video clip of this issue on those platforms and post to the youtube?

**Comment by deitch** [comment]
You are welcome to record, but it will not help debugging and resolving. Don't you work for arm? Can you ask them to make a server or two available for a few days, we can VPN in and then you can wipe it clean?

I'm asking some other routes, but not relying on it.

**Comment by rn** [comment]
we can probably get access to qualcomm via packet.net. need to send emails. We have a Huawei system in our packet.net account but it is unresponsive on the network (rebooted several times and the console logs look ok but no ssh...)

**Comment by deitch** [comment]
> we can probably get access to qualcomm via packet.net. 

Their hardware or ours? If it is theirs, I can ask as well (actually just sent a message on a slack channel there). If it is ours, different story.

**Comment by arm64b** [comment]
> Don't you work for arm? Can you ask them to make a server or two available for a few days, we can VPN in and then you can wipe it clean?

Oops, those servers are connected to the internal networking, so I really don't think you can VPN in :)
If you can access those 2 servers in packet.net, I think Ed will be happy to be help, but IMO seems only Cavium servers in their cloud, maybe something changed and packet.net has Qualcomm and Huawei platforms?

**Comment by deitch** [comment]
Heh, I asked Ed. But I hope he is asleep now. 

They only advertise Cavium, but maybe they have access? 

Can you get 2 servers connected to the Net?

**Comment by arm64b** [comment]
> Can you get 2 servers connected to the Net?

You mean to the external internet? Ah, I can't since this is IT infrastructure and we also have NDA signed with Qualcomm. IOW, those servers can be used inside Arm only.

**Comment by arm64b** [comment]
I make a screen record, the first part is on Qualcomm platform, second part is on Cavium platform, but seems the resolution is not good :(

[https://www.youtube.com/watch?v=zXX1EXD-wQY&feature=youtu.be](https://www.youtube.com/watch?v=zXX1EXD-wQY&feature=youtu.be)

**Comment by arm64b** [comment]
Hmm, more clear video record by my cell phone. But I can use one hand to take the phone, another hand to make some operations, so it's not convenient for me to do that :)
[https://youtu.be/-94FmErlC-Q](https://youtu.be/-94FmErlC-Q)

**Comment by vielmetti** [comment]
Looking at this now @rn @arm64b @deitch .

Just to be clear, this is currently seen on Qualcomm and Huawei and not on Cavium ThunderX.

**Comment by rn** [comment]
Right, I can't repro on Cavium and looks like @arm64b can only repro on Huawei and Qualcom

**Comment by deitch** [comment]
I did manage to reproduce it on a Huawei. @arm64b did you get anywhere in some possible lower-level symptoms (let alone root causes)?

**Comment by rn** [comment]
@deitch how did you repro. can you provide exact commands and description of symptoms

**Comment by deitch** [comment]
Sure, although I cannot be sure that what I have is identical to what @arm64b has. Also, I tend in your direction that this is more of a qemu or similar issue than `lkt` per se, but...

1. Clean Huawei arm64. 
2. Install `go`, `git`, `qemu`, `docker` (the usual suspects)
3. Install `linuxkit` and `moby`
4. Create a simple `test.yml` (stripped down version of `linuxkit.yml` from the repo, without nginx - see below)
4. `moby build -fromat kernel+initrd test.yml` - this works fine

**Comment by deitch** [comment]
I assume the `tty` messages are because of:

```
  cmdline: "console=tty0 console=ttyS0 console=ttyAMA0"
```

which may not exist? Don't know.

**Comment by deitch** [comment]
FYI, nothing particularly chewing up CPU or mem (as in, we tried "the usual suspects").

**Comment by deitch** [comment]
FWIW, the first time it hangs in the slow instance is here:

```
[    3.833000] virtio-pci 0000:00:02.0: virtio_pci: leaving for legacy driver
[    3.835734] xenfs: not registering filesystem on non-xen platform
[    3.837170] Serial: 8250/16550 driver, 4 ports, IRQ sharing enabled
```

**Comment by deitch** [comment]
Also, I killed both, cleaned up state, then ran in reverse order. Same thing: whichever starts first, runs; whichever does not, hangs.

**Comment by deitch** [comment]
Ah yes. Also, once the second hangs, exiting the first does not help.

**Comment by deitch** [comment]
And... I can get _both_ to hang by:

1. Start the first - it works
2. Start the second - it hangs
3. `halt` the first - it exits
4. Restart the first - it hangs too

**Comment by rn** [comment]
@deitch This is not a chat and you can edit existing comments (it's really hard to follow many piece by piece comments). I'm looking for a steps to reproduce (commands executed), which kernel version you running this on, version of qemu, which EFI firmware used, etc in a *single* comment.

**Comment by deitch** [comment]
Yeah, I know. Try switching back and forth between Slack (where "Enter" = submit, "Cmd-Enter" = newline) and GitHub (exact opposite). I kept submitting by accident as I swapped between them. Sorry about that.

**Comment by rn** [comment]
I got access to a HiSilicon machine on packet.net (thanks @vielmetti !).

Machines:
- `linuxkit-dev-arm64`: 96 core Cavium ThunderX CN8890, 4.10.0-26-generic, qemu 2.5.0
- `linuxkit-hisilicon-rneugeba`: 64 core HiSilicon Hi1616, 4.12.13, qemu 2.5.0

I also tried to install a 4.12.13 kernel on A Cavium ThunderX but failed...

## kernel+initrd testing

We use the LinuxKit `minimal.yml` image for testing build with `moby
build minimal.yml` build from `8f12286b77f8`. This uses the
`4.9.60-linuxkit` kernel.

We use this command to run it:

```
qemu-system-aarch64 -smp 1 -m 1024 -nographic \
    -cpu host -enable-kvm -machine virt,gic_version=host \
    -device virtio-rng-pci \
    -device virtio-net-pci,netdev=t0 -netdev user,id=t0 \
    -kernel minimal-kernel -initrd minimal-initrd.img \
    -append console=ttyAMA0
```

### HiSilicon

- First VM boots in ~3.6 second to `init` with the largest chunk of
  time spent in decompressing the `initrd`.
- While first VM is running, the second VM takes **168.88s** to boot
  to init (note, this time varies, it can be as "low" as 30s).
  
Looking at the boot log we see in the second VM:
```
[    3.548698] pci 0000:00:02.0: BAR 0: assigned [io  0x1020-0x103f]
[    3.556252] virtio-pci 0000:00:01.0: enabling device (0000 -> 0001)
[    3.558017] virtio-pci 0000:00:01.0: virtio_pci: leaving for legacy driver
[    3.560965] virtio-pci 0000:00:02.0: enabling device (0000 -> 0001)
[    3.562334] virtio-pci 0000:00:02.0: virtio_pci: leaving for legacy driver
[    3.564971] xenfs: not registering filesystem on non-xen platform
[    3.566311] Serial: 8250/16550 driver, 4 ports, IRQ sharing enabled
[  168.622874] Unable to detect cache hierarchy from DT for CPU 0
[  168.662155] loop: module loaded
[  168.741215] nbd: registered device at major 43
[  168.747113] megasas: 06.811.02.00-rc1
```
where all the time it spent.

I also tried without the `gic_version=host`, which was required on
Cavium, but this did not make a difference.

I also tried running without KVM, but this command just hangs (need to try again):
```
qemu-system-aarch64 -smp 1 -m 1024 -nographic \
    -cpu cortex-a57 -machine virt \
    -device virtio-rng-pci \
    -device virtio-net-pci,netdev=t0 -netdev user,id=t0 \
    -kernel minimal-kernel -initrd minimal-initrd.img \
    -append console=ttyAMA0
```

### Cavium

Both VMs boot around the same speed, no long delay for the second VM. 


## EFI Shell (qemu-efi)

@arm64b was using EFI booting, but instead of trying to boot an EFI
ISO, we just boot into the EFI shell as this is simpler.

For some reason I get an `EPERM` when using `/usr/share/qemu-efi/QEMU_EFI.fd`, and if I copy QEMU_EFI.fd is get a different error, so I did:
```
sudo apt-get install qemu-efi
dd if=/dev/zero of=flash0.img bs=1M count=64
dd if=/usr/share/qemu-efi/QEMU_EFI.fd of=flash0.img conv=notrunc
```

Then booting into the EFI shell with:
```
qemu-system-aarch64 -smp 1 -m 1024 -nographic \
    -cpu host -enable-kvm -machine virt \
    -drive if=pflash,format=raw,file=./flash0.img
```

Running this simultaneously on the HiSilicon or the Cavium shows no
delays.


## EFI ISO

Build an iso with `moby build -format iso-efi minimal.yml` and then ran it with (`flash0.img` is the one created above):
```
qemu-system-aarch64 -smp 1 -m 1024 -nographic \
    -cpu host -enable-kvm -machine virt,gic_version=host \
    -device virtio-rng-pci \
    -device virtio-net-pci,netdev=t0 -netdev user,id=t0 \
    -boot d -cdrom minimal-efi.iso \
    -drive if=pflash,format=raw,file=./flash0.img
```

### HiSilicon

The first VM boots fine, but then I boot the second VM (while the first on is still running) it hangs at:
```
  Booting `LinuxKit ISO Image'

EFI stub: Booting Linux Kernel...
```
for more than 10 minutes.

Running `sudo strace -ftp <pid>` on the hanging process gives us:
```
[...]
[pid 12053] 13:04:49 ioctl(15, KVM_RUN, 0) = 0
[pid 12053] 13:04:49 ioctl(15, KVM_RUN, 0) = 0
[pid 12053] 13:04:49 ioctl(15, KVM_RUN, 0) = 0
[pid 12053] 13:04:49 ioctl(15, KVM_RUN, 0) = 0
[pid 12053] 13:04:49 ioctl(15, KVM_RUN, 0) = 0
[pid 12053] 13:04:49 ioctl(15, KVM_RUN, 0) = 0
[pid 12053] 13:04:49 ioctl(15, KVM_RUN, 0) = 0
[pid 12053] 13:04:49 ioctl(15, KVM_RUN, 0) = 0
[pid 12053] 13:04:49 ioctl(15, KVM_RUN <unfinished ...>
[pid 12049] 13:04:49 <... ppoll resumed> ) = 0 (Timeout)
[pid 12049] 13:04:49 clock_gettime(CLOCK_MONOTONIC, {129144, 94340729}) = 0
[pid 12049] 13:04:49 clock_gettime(CLOCK_MONOTONIC, {129144, 94383589}) = 0
[pid 12049] 13:04:49 clock_gettime(CLOCK_MONOTONIC, {129144, 94424729}) = 0
[pid 12049] 13:04:49 clock_gettime(CLOCK_MONOTONIC, {129144, 94466009}) = 0
[pid 12049] 13:04:49 ppoll([{fd=0, events=POLLIN}, {fd=4, events=POLLIN}, {fd=7, events=POLLIN}, {fd=9, events=POLLIN}, {fd=10, events=POLLIN}, {fd=17, events=POLLIN}, {fd=18, events=POLLIN}], 7, {1, 0}, NULL, 8 <unfinished ...>
[pid 12053] 13:04:49 <... ioctl resumed> , 0) = 0
[pid 12053] 13:04:49 ioctl(15, KVM_RUN, 0) = 0
[pid 12053] 13:04:49 ioctl(15, KVM_RUN, 0) = 0
[pid 12053] 13:04:49 ioctl(15, KVM_RUN, 0) = 0
[pid 12053] 13:04:49 ioctl(15, KVM_RUN, 0) = 0k
[...]
```

On a second attempt, even running just one VM has delays like:
```
[    0.522148] virtio-pci 0000:00:02.0: virtio_pci: leaving for legacy driver
[    0.523733] virtio-pci 0000:00:03.0: virtio_pci: leaving for legacy driver
[    0.525295] xenfs: not registering filesystem on non-xen platform
[    0.526209] Serial: 8250/16550 driver, 4 ports, IRQ sharing enabled
[   55.224419] Unable to detect cache hierarchy from DT for CPU 0
[   55.263258] loop: module loaded
[   55.340248] nbd: registered device at major 43
[   55.356192] megasas: 06.811.02.00-rc1
[   55.356821] hisi_sas: driver version v1.6
```

Then starting a second VM hangs again at the EFI Stub as above.

Killing the first VM seems to **eventually** unlock the second VM (after several minutes). First we see:
```
  Booting `LinuxKit ISO Image'

EFI stub: Booting Linux Kernel...
EFI stub: Using DTB from configuration table
EFI stub: Exiting boot services and installing virtual address map...
```
then after another minute or so the normal boot resumes.


### Cavium

This works fine with two concurrent VMs.


## Summary/Next Step

Given that we use different kernel in the Cavium and the HiSilicon servers we should first  try repeating the experiments with the exact same kernels. This is probably best done by trying to boot a LinuxKit Image with QEMU and the minimal kernel/initrd/iso embedded into the image.

The hang when booting the second ISO based VM could be related to the long delay observed when booting `kernel+initrd`.

**Comment by arm64b** [comment]
Hello guys, according to the perf record information and strace, I think this is probably related to interaction between QEMU and UEFI FW on AArch64... As I said early, this issue doesn't show on x86_64 with same operation command and procedure...

**Comment by deitch** [comment]
>  I think this is probably related to interaction between QEMU and UEFI FW on AArch64.

@arm64b didn't we see the same issue even when booting direct with `kernel+initrd`, i.e. without firmware? Or did you mean the firmware on the underlying system? And if so, how?

> Huawei has its private kernel, so can't apply to the latest upstream kernel version

Do we know how it is different?

**Comment by arm64b** [comment]
Hello @deitch, if we don't specify the UEFI FW when booting the `kernel+initrd` format, the issue is not there. Because we are working on the QEMU VM, so when I say UEFI FW, it's the emulated UEFI for the VM just we have a UEFI installed on a bare metal machine which is mandatory to boot the UEFI-dependent image like uefi-iso or uefi-raw.

**Comment by deitch** [comment]
> when booting the kernel+initrd format, the issue is not there

I thought it was? Both my tests and @rn's indicated it was.

**Comment by rn** [comment]
I can confirm that on HiSilicon with `-device virtio-rng-pci` removed and two concurrent VMs started with:
```
qemu-system-aarch64 -smp 1 -m 1024 -nographic \
    -cpu host -enable-kvm -machine virt,gic_version=host \
    -device virtio-net-pci,netdev=t0 -netdev user,id=t0 \
    -kernel minimal-kernel -initrd minimal-initrd.img \
    -append console=ttyAMA0
```
there is no delay for the second VM

**Comment by rn** [comment]
To summarise some discussion from #2685. The issue seems to be that the virtio RNG device defaults to `/dev/random` and that blocks if there is not enough entropy (possibly due to lack of suitable hardware drivers being enabled/present). @ijc suggested using `/dev/urandom` instead and testing with modified example [above](https://github.com/linuxkit/linuxkit/issues/2636#issuecomment-341972998):
```
qemu-system-aarch64 -smp 1 -m 1024 -nographic \
    -cpu host -enable-kvm -machine virt,gic_version=host  \
    -object rng-random,filename=/dev/urandom,id=rng0   -device virtio-rng-pci,rng=rng0 \
    -device virtio-net-pci,netdev=t0 -netdev user,id=t0 \
    -kernel minimal-kernel -initrd minimal-initrd.img \
    -append console=ttyAMA0
```
seems to stop the stall when starting the second VM on the HiSilicon machine

The same applies to the EFI ISO boot with:
```
qemu-system-aarch64 -smp 1 -m 1024 -nographic \
    -cpu host -enable-kvm -machine virt,gic_version=host \
    -object rng-random,filename=/dev/urandom,id=rng0   -device virtio-rng-pci,rng=rng0 \
    -device virtio-net-pci,netdev=t0 -netdev user,id=t0 \
    -boot d -cdrom minimal-efi.iso \
    -drive if=pflash,format=raw,file=./flash0.img
```

reading `head -1 /dev/random` inside the VM then also does not block, which it did when doing so with `/dev/random` backed `virtio-rng-pci`
